File: backend/app/schemas/lesson.py
from datetime import datetime
from typing import TYPE_CHECKING, Literal

from pydantic import BaseModel, Field

if TYPE_CHECKING:
    from app.schemas.review import ReviewResponse


class LessonExerciseInput(BaseModel):
    exercise_id: str
    order_index: int = Field(ge=0)
    section: Literal["warmup", "main", "cooldown"]
    planned_duration_seconds: int | None = Field(default=None, gt=0)
    sets: int | None = Field(default=None, gt=0)
    reps: int | None = Field(default=None, gt=0)
    selected_modification: str | None = None
    instructor_notes: str | None = None


class LessonExerciseCompletionInput(BaseModel):
    id: str
    completion_status: Literal["completed", "shortened", "skipped"]
    actual_duration_seconds: int | None = Field(default=None, gt=0)


class LessonExerciseResponse(BaseModel):
    id: str
    lesson_id: str
    exercise_id: str
    order_index: int
    section: str
    planned_duration_seconds: int | None = None
    actual_duration_seconds: int | None = None
    sets: int | None = None
    reps: int | None = None
    selected_modification: str | None = None
    instructor_notes: str | None = None
    completion_status: str | None = None


class AttendanceMemberSummary(BaseModel):
    id: str
    name: str
    is_active: bool


class AttendanceResponse(BaseModel):
    id: str
    lesson_id: str
    group_member_id: str
    registered: bool
    attended: bool
    member: AttendanceMemberSummary | None = None


class CreateLessonRequest(BaseModel):
    title: str = Field(min_length=1)
    primary_goal: str | None = None
    secondary_goals: list[str] = Field(default_factory=list)
    level: Literal["beginner", "intermediate", "advanced"] | None = None
    planned_duration_minutes: int | None = Field(default=None, gt=0)
    instructor_notes: str | None = None
    lesson_exercises: list[LessonExerciseInput] = Field(default_factory=list)


class UpdateLessonRequest(BaseModel):
    title: str | None = Field(default=None, min_length=1)
    status: Literal["draft", "planned", "taught", "cancelled"] | None = None
    scheduled_for: datetime | None = None
    primary_goal: str | None = None
    secondary_goals: list[str] | None = None
    level: Literal["beginner", "intermediate", "advanced"] | None = None
    planned_duration_minutes: int | None = Field(default=None, gt=0)
    actual_duration_minutes: int | None = Field(default=None, gt=0)
    instructor_notes: str | None = None
    lesson_exercises: list[LessonExerciseInput] | None = None
    exercise_completions: list[LessonExerciseCompletionInput] | None = None


class LessonSummary(BaseModel):
    id: str
    group_id: str
    title: str
    status: str
    scheduled_for: datetime | None = None
    created_at: datetime
    updated_at: datetime
    has_review: bool


class LessonListResponse(BaseModel):
    lessons: list[LessonSummary]
    count: int


class LessonResponse(BaseModel):
    id: str
    group_id: str
    title: str
    status: str
    scheduled_for: datetime | None = None
    primary_goal: str | None = None
    secondary_goals: list[str]
    level: str
    planned_duration_minutes: int | None = None
    actual_duration_minutes: int | None = None
    instructor_notes: str | None = None
    created_at: datetime
    updated_at: datetime
    lesson_exercises: list[LessonExerciseResponse]
    attendance: list[AttendanceResponse] = Field(default_factory=list)
    review: "ReviewResponse | None" = None


class AttendanceUpdateItem(BaseModel):
    member_id: str
    registered: bool | None = None
    attended: bool | None = None


class AttendanceUpdateRequest(BaseModel):
    attendance: list[AttendanceUpdateItem]


class AttendanceListResponse(BaseModel):
    attendance: list[AttendanceResponse]
    count: int


class ReplaceExerciseRequest(BaseModel):
    new_exercise_id: str


from app.schemas.review import ReviewResponse

LessonResponse.model_rebuild()

File: mobile/src/types/lesson.ts
export type LessonSection = 'warmup' | 'main' | 'cooldown';

export type LessonStatus = 'upcoming' | 'completed' | 'cancelled';

export interface BuilderExercise {
  localId: string;
  exerciseId: string;
  nameHe: string;
  section: LessonSection;
}

export interface LessonExerciseInput {
  exercise_id: string;
  order_index: number;
  section: LessonSection;
  planned_duration_seconds?: number;
  sets?: number;
  reps?: number;
  selected_modification?: string;
  instructor_notes?: string;
}

export interface CreateLessonInput {
  title: string;
  primary_goal?: string;
  secondary_goals?: string[];
  instructor_notes?: string;
  planned_duration_minutes?: number;
  scheduled_for?: string;
  lesson_exercises: LessonExerciseInput[];
}

export interface UpdateLessonInput {
  title?: string;
  status?: LessonStatus;
  scheduled_for?: string | null;
  primary_goal?: string | null;
  instructor_notes?: string | null;
  planned_duration_minutes?: number;
  lesson_exercises?: LessonExerciseInput[];
}

export interface LessonSummary {
  id: string;
  group_id: string;
  title: string;
  status: LessonStatus;
  scheduled_for: string | null;
  planned_duration_minutes: number | null;
  created_at: string;
  updated_at: string;
  has_review: boolean;
}

export interface LessonExercise {
  id: string;
  lesson_id: string;
  exercise_id: string;
  order_index: number;
  section: LessonSection;
  planned_duration_seconds: number | null;
  actual_duration_seconds: number | null;
  sets: number | null;
  reps: number | null;
  selected_modification: string | null;
  instructor_notes: string | null;
  completion_status: string | null;
}

export interface Lesson {
  id: string;
  group_id: string;
  title: string;
  status: LessonStatus;
  scheduled_for: string | null;
  primary_goal: string | null;
  secondary_goals: string[];
  level: string;
  planned_duration_minutes: number | null;
  actual_duration_minutes: number | null;
  instructor_notes: string | null;
  created_at: string;
  updated_at: string;
  lesson_exercises: LessonExercise[];
}
